### ISSUE TITLE
Consider exposing supported syntax and its variants

### ISSUE THREAD

**Issue by bebraw** [lukejacksonn/oceanwind#9]
I use TypeScript to provide typing over Oceanwind in my current work. I provide a `sx` prop for components where you can provide Oceanwind classes within a string (could be an array of strings).

At the moment, the type for this is `any` and that makes `sx` a little dangerous to use. To improve DX, it would be better to allow only syntax that's supported by Oceanwind.

Would it be feasible to generate the syntax and its different permutations within a module so that it's possible to type systems to infer available options? I am currently using this kind of approach against the theme for type inference for more specific props than `sx`.

**Comment by lukejacksonn** [comment]
This kind of developer experience sounds great and I'd certainly like to help facilitate generating the syntax file. I don't know exactly what format you require it in but the grammar file that the switch statement was initially generated with will probably be of use here regardless:

https://gist.github.com/kitten/219ddda1db5df4ad42a05abf0f2738dd#file-grammar-js

In fact @kitten is the most likely person to know if this is doable or not and how far off we are already. Might be worth reaching out to him if you are interested.

**Comment by kitten** [comment]
I think the trickiest part is generating types that work with the theme file. I know for a fact that it will be possible once Template Literal types Land in TypeScript: https://github.com/microsoft/TypeScript/pull/40336#issuecomment-690709136

**Comment by bebraw** [comment]
To give you a better idea, what I'm looking to do is to replace sx at my [Box component](https://github.com/survivejs/tailwind-webpack-starter/blob/feat-deno/ds/primitives/box.tsx) with a better type. It would be enough to enumerate all the possible classes and their combinations there.

In the current approach I infer a lot based on the theme and I believe a similar way could work for typing. In essence, if there was something like `type OceanwindClasses = 'flex' | 'flex-grow' | ...`, that would be perfect. That's likely a composite so maybe you get `type Padding = 'p-2' | 'p-4' | ...` for each main feature and then compose the bigger type out of these.

Of course where this gets interesting is with all the possible variants since we can have `hover:`, `sm:`, and other modifiers but maybe there's a cool way to pull that off in the type system.

I agree template literate types may help a lot here. In some way the problem is about mapping the class grammar to the TypeScript type system and if we can achieve that, I believe it's great for developer experience and for various use cases.

**Comment by lukejacksonn** [comment]
Humm, I see. Well generating a big long string of possible options certainly isn't impossible, an exhaustive list would be _huge_ though. A good majority of unique grammars are mapped out in the [module tests file](https://github.com/lukejacksonn/oceanwind/blob/master/test/module.js) but it doesn't go so far as to test every color variant on all directives that accept a color even. That would be a _lot_ already, and then it has to account for cases that extend theme.

So my initial thoughts are it would need to be smarter than a big long list of types. Perhaps that is where Template Literal types come in. But honestly I'm not a typescript expert! Would still love to see something like this working. I believe most of what is required is actually here but piecing it altogether could take a while 😅

**Comment by bebraw** [comment]
@lukejacksonn Yeah, I see what you are saying. A big long list of exhaustive options would be too much. It definitely looks like a grammar is needed for tackling this.

Does Tailwind have an explicit grammar or is it all implicit and within their PostCSS plugin definition?

Maybe it's good to revisit this issue when TypeScript gains improvements on this department.

**Comment by kitten** [comment]
@bebraw we already have a full grammar of Tailwind since part of the transformer to CSS was auto generated originally.

This could also generate the full things which isn't an issue at all (it also doesn't matter if they get big if it's generated)

However part of the theme and grammar is really dynamic depending on the theme.

**Comment by bebraw** [comment]
> we already have a full grammar of Tailwind since part of the transformer to CSS was auto generated originally.

Ah, cool. Do you have the generator code somewhere public?

> However part of the theme and grammar is really dynamic depending on the theme.

That's true. I was thinking maybe I could try to use your generator to generate some kind of type(s) while taking that into account. 😎

**Comment by kitten** [comment]
@bebraw certainly! The original code is here: https://gist.github.com/kitten/219ddda1db5df4ad42a05abf0f2738dd

**Comment by bebraw** [comment]
Awesome, thanks a lot.

**Comment by bebraw** [comment]
Now that TypeScript 4.1 is out, people are starting to type Tailwind. Example: https://twitter.com/markdalgleish/status/1329639099498577921 . [Demo](https://www.typescriptlang.org/play?ts=4.1.0-beta#code/C4TwDgpgBA6glsAFgIQE4QIYGswHs4B2wAzgDwAaUEAHsBAQCbFTHCqEDmAfFALxQADACQBvAOTEAtmKgAfKGMkMZ8sQBsOKhdTVaxAJh1iAvgC5R5YwICwAKFCQoyDMQgBhNS9fN+YgGa4uHoARhioIRgAXmIA3FAA9PFQACqIEK5QYdC4wQBucLgArsRqIFAEuMBQ6BhqKRhwagDuhAxQAMaexN4ANFDBhVVqEMASUGDodIx2DtAAarVwDB5e6XxOLu5d3nKwCCg1OPhEZM6uK93pXDPg84vL2xQ8-JQ0U0xQC2pLFzsA-FBKKZyhBchBUDE7DdHMkGs1WgBRVCoXCoUjJKi0egfVjsAjcdYiKjI1HAtiFaDGKAAMhSUPst3qjRajF+6VIbkx72YuM4fQAysAMHQudieWxOOsxGJnnYoPKoJy3mLBKIvj9HoQ-OCoAAxOCoVhcKmiLU6gBK6WAVjlCrtANhzNabLIltYfWEIkFwugysYzGltrtwagAMDthDIeBnu9dCpAmMavuv1I+sNwGNAmuEcjwKVWP9n2T23SQcj8oBMaFIr9H2loYUMmjoljEHjiZEbhtOfL8uBjvhjCRKLRAgAkgRcvcmYO2p0vMDRPnuaqRGbUHqDawTWuCNqN27rQIG2nWFA81YuJDbHY-IUCO1gAUCFBgE0OaLC7z8VwABTz7oADkMEkdJ+zhFkHlWMg3C4ABKKAiUmQpUBfADiGA0DmGMelEigAAZEYxjYMoECgIpgFMek31-fxAjEODqKaWiAlwfowgYpiWMCdiN1CaJGJvexmLotiAHkJIkgAFfleM4oS8MA3AmigFokH6Q48EIEgqKEmiJEkUxWKgJRTFCDcNDMqJ5JmESpCMnjTMk6TZMs-j5KAA).

**Comment by kitten** [comment]
Agreed that it's the perfect time to trial this ✌️
